---
title: "Human whole muscle vs mitochondrial miRNA Analysis"
author: "Jessica Silver (additional analysis by Megan Soria)"
date: 2023-07-09
output:
  html_document:
    theme: lumen
    toc: yes
    toc_float: true
  html_notebook:
    theme: lumen
    toc: yes
    toc_float: true
  prettydoc::html_pretty:
    theme: hpstr
    highlight: github
---

source: https://github.com/jessicasilver/miRNAs

# Summary and background information
The data used in this analysis is from an exercise intervention study in 12 human participants conducted by **[name]** on **[year]**. Muscle biopsies were taken at 3 time points for each participant: pre-exercise, post-exercise, and 3 hours post-exercise <br> <br>
A whole muscle and a mitochondrial fraction is extracted from each biopsy. <br> <br>

This analysis focuses on the differential expression of the miRNA of human whole muscle and mitochondrial transcriptome.  

# Load and install packages
The following packages are required for the following analysis and visualisation. The code below will automatically install and load packages from CRAN. Packages from Bioconductor can be installed by un-commenting (removing the # before the line or selecting the code block and pressing ctrl+shift+c) and running  the code block for the Bioconductor packages. Sometimes these installations will throw an error and specific packages might need to be installed manually.

```{r echo = T, results = 'hide'}
suppressPackageStartupMessages({c(
  # packages from CRAN
  if(!require("prettydoc")) install.packages("prettydoc"), library(prettydoc),
  if(!require("readxl")) install.packages("readxl"), library(readxl),
  if(!require("dplyr")) install.packages("dplyr"), library(dplyr),
  if(!require("RColorBrewer")) install.packages("RColorBrewer"), library(RColorBrewer),
  if(!require("circlize")) install.packages("circlize"), library(circlize),
  library(ggplot2),
  library(RNAseqQC),
  # packages from Bioconductor
  # run the installation code when needed
  # if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")
  #   BiocManager::install(c("DESeq2", "ComplexHeatmap", "EnhancedVolcano", "org.Hs.eg.db")),
  library(DESeq2),
  library(ComplexHeatmap),
  library(EnhancedVolcano),
  library(cowplot),
  library(PCAtools),
  library(fgsea),
  library(msigdbr),
  library(tidyverse),
  library(dplyr),
  library(biomaRt),
  library(clusterProfiler),
  library(enrichplot)
  )})

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("DESeq2")
  
# load self-authored functions 
source("functions.R")
```

***
# Load data for skeletal muscle
```{r}
muscle.cts <- read.csv("data/silver_muscle raw.csv", sep=",", row.names = 1) #keeps miR identifiers as rownames
muscle.cts[is.na(muscle.cts)] <-0 #replace NA with 0 #now a num not int
muscle.cts <- as.matrix(muscle.cts) #convert to matrix

#read in column datafile aka sample information file
muscle.coldata <- read.csv("data/silver_muscle condition.csv", sep = ",", header = TRUE) # do not incl row.names=1

muscle.coldata <- muscle.coldata[,c("sampleid","condition")]
muscle.coldata$condition <- factor(muscle.coldata$condition)
muscle.coldata$sampleid <- factor(muscle.coldata$sampleid)

all(rownames(muscle.coldata$sampleid) == colnames(muscle.cts)) #check these both come back as TRUE
```

# Load data for mitochondria
```{r}
mito.cts <- read.csv("data/silver_hmt raw_by group.csv", sep=",", row.names = 1) #keeps miR identifiers as rownames
mito.cts[is.na(mito.cts)] <-0 #replace NA with 0 #now a num not int
mito.cts <- as.matrix(mito.cts) #convert to matrix

#read in column datafile aka sample information file
mito.coldata <- read.csv("data/silver_hmt condition_by group.csv", sep = ",", header = TRUE) # do not incl row.names=1

mito.coldata <- mito.coldata[,c("sampleid","condition")]
mito.coldata$condition <- factor(mito.coldata$condition)
mito.coldata$sampleid <- factor(mito.coldata$sampleid)

all(rownames(mito.coldata$sampleid) == colnames(mito.cts)) #check these both come back as TRUE



mito.cts <- read.csv("data/silver_hmt raw_by group.csv", sep=",", row.names = 1) #keeps miR identifiers as rownames
mito.cts[is.na(mito.cts)] <-0 #replace NA with 0 #now a num not int
mito.cts <- as.matrix(mito.cts) #convert to matrix

#read in column datafile aka sample information file
mito.coldata <- read.csv("data/silver_hmt condition_by group.csv", sep = ",", header = TRUE) # do not incl row.names=1

mito.coldata <- mito.coldata[,c("sampleid","condition")]
mito.coldata$condition <- factor(mito.coldata$condition)
mito.coldata$sampleid <- factor(mito.coldata$sampleid)

all(rownames(mito.coldata$sampleid) %in% colnames(mito.cts)) #checking that count matrix and coldata are consistent in terms of sample order
all(rownames(mito.coldata$sampleid) == colnames(mito.cts)) #check these both come back as TRUE
```

# Differential expression analysis 
The following differential expression analysis is conducted using the DESeq2 package

## Differential expression anlysis for muscle
```{r}
#set up deseq2 object
dds <- DESeqDataSetFromMatrix(countData = muscle.cts,
                              colData = muscle.coldata,
                              design = ~ condition)
x <- dim(dds)

dds$condition <- relevel(dds$condition, ref = "PRE") #set your baseline. #hsa=PRE #rno=SED

keep <- rowSums(counts(dds)) >=10  
dds <- dds[keep,]

cat("Before filtering for lowcounts: ", x, "\n",
    "After filtering for lowcounts: ", dim(dds))

dds <- DESeq(dds) #run deseq2

# normalise: use varianceStabilizingTransformation() for nrow(dds) < 1000
muscle.vst <- varianceStabilizingTransformation(dds, blind=F) 

#extract the results for each comparison that you are interested in 
muscle.res <- as.data.frame(results(dds, contrast=c("condition","3HPOST","PRE")))
muscle.res <- muscle.res[order(muscle.res$pvalue),]
summary(muscle.res)
head(muscle.res)
```

## Differential expression anlysis for mitochondria
```{r}

#set up deseq2 object
dds <- DESeqDataSetFromMatrix(countData = mito.cts,
                              colData = mito.coldata,
                              design = ~ condition)

dds$condition <- relevel(dds$condition, ref = "PRE") #set your baseline. #hsa=PRE #rno=SED

keep <- rowSums(counts(dds)) >= 10   #thesolding. first num = read counts, second num = x samples
#for DE seq on whole skeletal muscle small miRNAs, threshold as `keep <- rowSums(counts(dds)) >=10 
dds <- dds[keep,]

dds <- DESeq(dds) #run deseq2

# normalise: use varianceStabilizingTransformation() for nrow(dds2) < 1000
mito.vst <- varianceStabilizingTransformation(dds, blind=F) 

#extract the results for each comparison that you are interested in 
mito.res <- as.data.frame(results(dds, contrast=c("condition","3HPOST","PRE")))
mito.res <- mito.res[order(mito.res$pvalue),]
summary(mito.res)
head(mito.res)
```
# Visualisations

# Density plot for gene counts of mitochondria vs muscle miRNAs
```{r}
# muscle
select2 <- as.matrix(assay(muscle.vst)) #scale ntd

#pull out samples by group or time-point
PRE <- select2[,1:9]
POST_0h <- select2[,10:21]
POST_3h <- select2[,22:33]

#row (i.e. miRNA) means by group or time-point
rmeansPRE <- as.vector(rowMeans(PRE))
rmeansPOST_0h <- as.vector(rowMeans(POST_0h))
rmeansPOST_3h <- as.vector(rowMeans(POST_3h))

df.muscle <- data.frame(rmeansPRE, rmeansPOST_0h, rmeansPOST_3h) %>% 
  mutate(Sample = "Muscle") %>% 
  mutate(genes = row.names(select2))


# mitochondria
select2 <- as.matrix(assay(mito.vst)) #scale ntd

#pull out samples by group or time-point
PRE <- select2[,1:9]
POST_0h <- select2[,10:21]
POST_3h <- select2[,22:33]

#row (i.e. miRNA) means by group or time-point
rmeansPRE <- as.vector(rowMeans(PRE))
rmeansPOST_0h <- as.vector(rowMeans(POST_0h))
rmeansPOST_3h <- as.vector(rowMeans(POST_3h))

df.mito <- data.frame(rmeansPRE, rmeansPOST_0h, rmeansPOST_3h) %>% 
  mutate(Sample = "Mito") %>% 
  mutate(genes = row.names(select2))

p <- function(v) {
  if(length(v) == 2){
    f=paste(v[1], "&", v[2])
  }else(f=paste(v))
  return(f)
 }
# pre muscle and mito
PRE.combined <- df.mito %>% 
  select(rmeansPRE, Sample, genes) %>%
  bind_rows(select(df.muscle, rmeansPRE, Sample, genes)) %>%
  group_by(genes) %>%
  summarise(rmeansPRE = mean(rmeansPRE),
            Sample = p(as.character(Sample)))

# 3hr post muscle and mito
POST3h.combined <- df.mito %>% 
  select(rmeansPOST_3h, Sample, genes) %>%
  bind_rows(select(df.muscle, rmeansPOST_3h, Sample, genes)) %>%
  group_by(genes) %>%
  summarise(rmeansPOST_3h = mean(rmeansPOST_3h),
            Sample = p(as.character(Sample)))

# post muscle and mito
POST.combined <- df.mito %>% 
  select(rmeansPOST_0h, Sample, genes) %>%
  bind_rows(select(df.muscle, rmeansPOST_0h, Sample, genes)) %>%
  group_by(genes) %>%
  summarise(rmeansPOST_0h = mean(rmeansPOST_0h),
            Sample = p(as.character(Sample)))

# plot
theme_set(theme_classic(base_size = 12))
ggplot(data=PRE.combined, aes(x=rmeansPRE, group=Sample, fill=Sample)) +
    geom_density(adjust=1.5, alpha=.6) +
  labs(title = "Normalised counts of miRNAs in humans:",
       subtitle = "mitochondira and muscle pre-exercise") +
    theme(plot.title=element_text(hjust=0.5),
          plot.subtitle=element_text(hjust=0.5)) 
  
ggplot(data=POST.combined, aes(x=rmeansPOST_0h, group=Sample, fill=Sample)) +
    geom_density(adjust=1.5, alpha=.6) +
    labs(title = "Normalised counts of miRNAs in humans:",
       subtitle = "mitochondira and muscle post-exercise") +
     theme(plot.title=element_text(hjust=0.5),
          plot.subtitle=element_text(hjust=0.5))

ggplot(data=POST3h.combined, aes(x=rmeansPOST_3h, group=Sample, fill=Sample)) +
    geom_density(adjust=1.5, alpha=.6) +
    labs(title = "Normalised counts of miRNAs in humans:",
       subtitle = "mitochondira and muscle 3 hours post-exercise") +
     theme(plot.title=element_text(hjust=0.5),
          plot.subtitle=element_text(hjust=0.5))

```


## PCA plots
```{r}
pca.muscle <- plot_pca(muscle.vst, PC_x = 1, PC_y = 2, color_by = "condition", point_rel_size = 5)
pca.mito <- plot_pca(mito.vst, PC_x = 1, PC_y = 2, color_by = "condition", point_rel_size = 5)
```


## Volcano plots
```{r}
# muscle
vplot.muscle <- EnhancedVolcano(as.data.frame(muscle.res), x="log2FoldChange", y="padj", lab = NA, 
                pCutoff = 0.05, FCcutoff = 1, title = "", subtitle = "Muscle miRNA DE Genes", gridlines.major = FALSE, gridlines.minor = FALSE, 
                axisLabSize = 14, titleLabSize = 0, subtitleLabSize = 16, captionLabSize = 14, legendLabSize = 14, legendIconSize = 4,
                col = c("grey30", "royalblue", "forestgreen", "red2"))

# mito
vplot.mito <- EnhancedVolcano(as.data.frame(mito.res), x="log2FoldChange", y="padj", lab = NA, 
                pCutoff = 0.05, FCcutoff = 1, title = "", subtitle = "Mitochondrial miRNA DE Genes", gridlines.major = FALSE, gridlines.minor = FALSE, 
                axisLabSize = 14, titleLabSize = 0, subtitleLabSize = 16, captionLabSize = 14, legendLabSize = 14, legendIconSize = 4,
                col = c("grey30", "royalblue", "forestgreen", "red2"))

vplot.muscle
vplot.mito

```


## Heatmaps
```{r fig.width=16.22}
# muscle
count_mat <- z_score(muscle.vst, muscle.res)
hmap_name <- "Muscle miRNAs"
hmap.muscle <- hmap(50, as.data.frame(muscle.res), count_mat, hmap_name)

# mito
count_mat <- z_score(mito.vst, mito.res)
hmap_name <- "Mitochondrial miRNAs"
hmap.mito <- hmap(50, as.data.frame(mito.res), count_mat, hmap_name)

hmap.muscle
hmap.mito
```


# Pathway analysis
## Convert ENSEMBL IDs
```{r}
# preVs3hrPost - muscle
muscle.ID <- muscle.res %>%
  rownames_to_column(., var = "miR_ID")

# preVs3hrPost - mito
mito.ID <- mito.res %>%
  rownames_to_column(., var = "miR_ID")

```

## Get gene set data 
Using the **msigdbr library** from the [Broad Institute's Molecular Signatures Database](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp).
```{r}
# get a list of all genes from the original DESeq db to add to the gene set data
# clusterProfiler has a tendency to "ignore" ids that does not have an annotation in the gene sets

# adds genes without annotation to the geneset
bg <- muscle.ID$miR_ID
bg <- unique(sapply(strsplit(bg," "),"[[",1))
bgdf <- data.frame("background",bg)
colnames(bgdf) <- c("gs_name","ensembl_gene")

# GO-Biological pathways
GO_BP <- msigdbr(species = "Homo sapiens", 
                 category = "C5", subcategory = "BP") %>% 
  dplyr::distinct(gs_name, ensembl_gene) %>% as.data.frame()

x <- dim(GO_BP)

GO_BP <- rbind(GO_BP,bgdf[!bgdf$ensembl_gene %in% GO_BP$ensembl_gene,]) %>% 
  mutate(gs_name = gsub("GOBP_", "", .$gs_name)) %>% 
  mutate(gs_name = gsub("_", " ", .$gs_name)) %>% 
           as.data.frame()

cat("Before adding background genes: ", x, "\n",
    "After adding background genes: ", dim(GO_BP), "\n",
    "Total added: ", dim(GO_BP) - x)
```

## Retrieve miRNA target regions via biomaRt
```{r}
# list of miRNA identifiers to convert
muscle_miRNA_ids <- unique(rownames(muscle.cts))

# show the list of databases from ENSEMBL
listEnsembl()

# from the list above, chose the 'regulation' database
ensembl_db <- useEnsembl(biomart = "regulation")
# show list of data sets from the chosen database
ensembl_datasets <- listDatasets(ensembl_db)

# from dataset above, choose "hsapiens_mirna_target_feature"
# connect the database and dataset
ensembl_conn <- useEnsembl(biomart = "regulation", dataset = "hsapiens_mirna_target_feature")

# query for gene names
# get list of attributes and filters to use in the query
attr <- listAttributes(ensembl_conn)
filters <- listFilters(ensembl_conn)

# build query from attributes and filters
muscle_miRNA_ensembl <- getBM(attributes = c("display_label", "gene_stable_id"), #this is what the output looks like
      filters = "external_identifier", #this is the format of our ensmbl ids
      values = muscle_miRNA_ids, # this is our original list
      mart = ensembl_conn)

# convert ensembl ids to symbols
muscle_ensembl_ids <- unique(muscle_miRNA_ensembl$gene_stable_id)

ensembl_conn2 <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
muscle_ensembl2GO <- getBM(attributes = c("ensembl_gene_id", "namespace_1003", "name_1006"), #this is what the output looks like
      filters = "ensembl_gene_id", #this is the format of our ensmbl ids
      values = muscle_ensembl_ids, # this is our original list
      mart = ensembl_conn2)

# Subset only biological processes
muscle_GBP <- muscle_ensembl2GO %>%
  dplyr::filter(.$namespace_1003 == "biological_process")
```


```{r}
# list of miRNA identifiers to convert
mito_miRNA_ids <- unique(rownames(mito.cts))

# show the list of databases from ENSEMBL
listEnsembl()

# from the list above, chose the 'regulation' database
ensembl_db <- useEnsembl(biomart = "regulation")
# show list of data sets from the chosen database
ensembl_datasets <- listDatasets(ensembl_db)

# from dataset above, choose "hsapiens_mirna_target_feature"
# connect the database and dataset
ensembl_conn <- useEnsembl(biomart = "regulation", dataset = "hsapiens_mirna_target_feature")

# query for gene names
# get list of attributes and filters to use in the query
attr <- listAttributes(ensembl_conn)
filters <- listFilters(ensembl_conn)

# build query from attributes and filters
mito_miRNA_ensembl <- getBM(attributes = c("display_label", "gene_stable_id"), #this is what the output looks like
      filters = "external_identifier", #this is the format of our ensmbl ids
      values = mito_miRNA_ids, # this is our original list
      mart = ensembl_conn)

# convert ensembl ids to symbols
mito_ensembl_ids <- unique(mito_miRNA_ensembl$gene_stable_id)

ensembl_conn2 <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
mito_ensembl2GO <- getBM(attributes = c("ensembl_gene_id", "namespace_1003", "name_1006"), #this is what the output looks like
      filters = "ensembl_gene_id", #this is the format of our ensmbl ids
      values = mito_ensembl_ids, # this is our original list
      mart = ensembl_conn2)

# Subset only biological processes
mito_GBP <- mito_ensembl2GO %>%
  dplyr::filter(.$namespace_1003 == "biological_process")
```

## Run ORA
### Set inputs
```{r}
# muscle
df = muscle.res 
sig_genes_df = subset(df, padj < 0.05)
genes <- sig_genes_df$log2FoldChange
names(genes) <- rownames(sig_genes_df)
genes <- na.omit(genes)
genes <- names(genes)[abs(genes) > 1]
  
# convert to ensembl
x <- getBM(attributes = c("display_label", "gene_stable_id"), 
      filters = "external_identifier", 
      values = genes, # this is our original list
      mart = ensembl_conn)
muscle_ORAgenes <- unique(x$gene_stable_id)

# mito
df = mito.res 
sig_genes_df = subset(df, padj < 0.05)
genes <- sig_genes_df$log2FoldChange
names(genes) <- rownames(sig_genes_df)
genes <- na.omit(genes)
genes <- names(genes)[abs(genes) > 1]
  
# convert to ensembl
x <- getBM(attributes = c("display_label", "gene_stable_id"), 
      filters = "external_identifier", 
      values = genes, # this is our original list
      mart = ensembl_conn)
mito_ORAgenes <- unique(x$gene_stable_id)

```

## run ORA from clusterProfiler
```{r}
library(org.Hs.eg.db)
go_enrich <- enrichGO(gene = muscle_ORAgenes,
                      universe = names(mito_ensembl_ids),
                      OrgDb = "org.Hs.eg.db", 
                      keyType = 'ENSEMBL',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.10)

upsetplot(go_enrich)
barplot(go_enrich, 
        drop = TRUE, 
        showCategory = 10, 
        title = "GO Biological Pathways",
        font.size = 8) 
dotplot(go_enrich)
#emapplot(pairwise_termsim(go_enrich))
```

# Comparison of enriched pathways