---
title: "Pathway Analysis for Whole Skeletal Muscle"
author: "MSoria"
date: '2023-05-31'
output:
  html_document:
    theme: lumen
    toc: yes
    toc_float: true
  html_notebook:
    theme: lumen
    toc: yes
    toc_float: true
  prettydoc::html_pretty:
    theme: hpstr
    highlight: github
---

# Purpose

This markdown file aims to perform a pathway analysis on the differentially expressed genes of an exercise intervention study.

### Install and load prerequisite R-packages

```{r load library, message=FALSE, warning=FALSE, include=FALSE, paged.print=TRUE}
# Installation of some useful packages
suppressPackageStartupMessages({c(
  if(!isTRUE(require("prettydoc"))){install.packages("prettydoc")},
  if(!isTRUE(require("readxl"))){install.packages("readxl")},
  library(prettydoc),
  library(readxl), # reads excel files
  library(here), # finds files based on current working directory
  library(DESeq2),
  library(biomaRt),
  library(tidyverse),
  library(dplyr),
  library(msigdbr),
  library(clusterProfiler),
  library(enrichplot),
  library(pathview)
  )})

# load self-authored functions 
source("functions.R")

```

# Source:
http://www.bioconductor.org/packages/devel/bioc/vignettes/rWikiPathways/inst/doc/Pathway-Analysis.html

# Load Differentially Expressed Genes
```{r}
ddsResults_list <- readRDS("/mnt/data/megan/Danielle_RA/Jess/miRNAs/Meg/rds/ddsResults_wholeMusc.rds")

# put contrasts into separate DBs and remove ensembl id versions 
preVs3hrPost <- ddsResults_list[[1]] %>% 
  rownames_to_column(var = "ENSEMBL") %>% 
  mutate(ENSEMBL = gsub("\\..*","", .$ENSEMBL))

preVpost <- ddsResults_list[[2]] %>% 
  rownames_to_column(var = "ENSEMBL") %>% 
  mutate(ENSEMBL = gsub("\\..*","", .$ENSEMBL))

postVs3hrPost <- ddsResults_list[[3]] %>% 
  rownames_to_column(var = "ENSEMBL") %>% 
  mutate(ENSEMBL = gsub("\\..*","", .$ENSEMBL))

# create clusterProfiler inputs for ORA and GSEA
preV3h_input <- cProfilr_input(preVs3hrPost)
preVpost_input <- cProfilr_input(preVpost)
postV3hP_input <- cProfilr_input(postVs3hrPost)

```


# Get gene set data 
Using the **msigdbr library** from the [Broad Institute's Molecular Signatures Database](https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp).

```{r echo=TRUE}
# get a list of all genes from the original DESeq db to add to the gene set data
# clusterProfiler has a tendency to "ignore" ids that does not have an annotation in the gene sets

# adds genes without annotation to the geneset
bg <- postVs3hrPost$ENSEMBL
bg <- unique(sapply(strsplit(bg," "),"[[",1))
bgdf <- data.frame("background",bg)
colnames(bgdf) <- c("gs_name","ensembl_gene")

# GO-Biological pathways
GO_BP <- msigdbr(species = "Homo sapiens", 
                 category = "C5", subcategory = "BP") %>% 
  dplyr::distinct(gs_name, ensembl_gene) %>% as.data.frame()

x <- dim(GO_BP)

GO_BP <- rbind(GO_BP,bgdf[!bgdf$ensembl_gene %in% GO_BP$ensembl_gene,]) %>% 
  mutate(gs_name = gsub("GOBP_", "", .$gs_name)) %>% 
  mutate(gs_name = gsub("_", " ", .$gs_name)) %>% 
           as.data.frame()

cat("Before adding background genes: ", x, "\n",
    "After adding background genes: ", dim(GO_BP), "\n",
    "Total added: ", dim(GO_BP) - x)
```

# Gene Set Enrichment analysis 
## Pre vs 3hr post 
```{r fig.height=10, fig.width=12, warning=FALSE}
preV3h_gseGO <- clusterProfiler::GSEA(geneList = preV3h_input$gene_list, 
                     TERM2GENE = GO_BP,
                     pvalueCutoff = 0.05,
                     minGSSize = 10, 
                     maxGSSize = 800,
                     eps = 0,
                     by = "fgsea",
                     verbose = FALSE)

# plot results
d1 <- dotplot(preV3h_gseGO, showCategory=10, split=".sign", 
        label_format = 35,
        font.size = 5,) +
        facet_grid(.~.sign) + 
        labs(title = "Gene Set Enrichment Analysis of DE Genes (3hr post vs Pre exercise)",
             subtitle = "FDR < 0.05, -1 > logfold2Change > 1 ") +
        theme_bw(base_size = 18) +
        theme(plot.title=element_text(hjust=0.5, size = 20, face = "bold"),
              plot.subtitle=element_text(vjust=2, hjust=0.5, size = 18,face = "bold"),
              axis.text.y = element_text(size = 11)) 
  

e1 <- emapplot(pairwise_termsim(preV3h_gseGO), showCategory = 15, layout = "circle") +
  ggtitle("Enrichment map plot of DE Genes (3hr post vs Pre exercise)") +
  theme_void(base_size = 18) +
  theme(plot.title=element_text(vjust=3, hjust=0.5, size = 20, face = "bold"))


r1 <- ridgeplot(preV3h_gseGO, showCategory = 15) + 
  labs(title = "Density plots of fold change frequency values per gene within each set",
       subtitle = "(3hr post vs Pre exercise)",
    x = "Enrichment Distribution") +
  theme_classic(base_size = 18) +
  theme(plot.title=element_text(hjust=0.5, size = 20, face = "bold"),
              plot.subtitle=element_text(hjust=0.5, size = 20, face = "bold"))

d1
e1
r1

pdf(file = "figures/wholeMusc_mRNA_preV3h_gseGO.pdf",   
    width = 12, # The width of the plot in inches
    height = 10) # The height of the plot in inches
d1
e1
r1
dev.off()
```
## Pre vs Post 
```{r fig.height=10, fig.width=12, warning=FALSE}
preVpost_gseGO <- clusterProfiler::GSEA(geneList = preVpost_input$gene_list, 
                     TERM2GENE = GO_BP,
                     pvalueCutoff = 0.05,
                     minGSSize = 10, 
                     maxGSSize = 800,
                     eps = 0,
                     by = "fgsea",
                     verbose = FALSE)

# plot results
d2 <- dotplot(preVpost_gseGO, showCategory=10, split=".sign", 
        label_format = 30,
        font.size = 5,) +
        facet_grid(.~.sign) + 
        labs(title = "Gene Set Enrichment Analysis of DE Genes (Post vs Pre exercise)",
             subtitle = "FDR < 0.05, -1 > logfold2Change > 1 ") +
        theme_bw(base_size = 18) +
        theme(plot.title=element_text(hjust=0.5, size = 20, face = "bold"),
              plot.subtitle=element_text(vjust=2, hjust=0.5, size = 18,face = "bold"),
              axis.text.y = element_text(size = 12)) 
  

e2 <- emapplot(pairwise_termsim(preVpost_gseGO), showCategory = 15, layout = "circle") +
  ggtitle("Enrichment map plot of DE Genes (Post vs Pre exercise)") +
  theme_void(base_size = 18) +
  theme(plot.title=element_text(vjust=3, hjust=0.5, size = 20, face = "bold"))


r2 <- ridgeplot(preVpost_gseGO, showCategory = 15) + 
  labs(title = "Density plots of fold change frequency values per gene within each set",
       subtitle = "(Post vs Pre exercise)",
    x = "Enrichment Distribution") +
  theme_classic(base_size = 18) +
  theme(plot.title=element_text(hjust=0.5, size = 20, face = "bold"),
              plot.subtitle=element_text(hjust=0.5, size = 20, face = "bold"))

d2
e2
r2

pdf(file = "figures/wholeMusc_mRNA_preVpost_gseGO.pdf",   
    width = 12, # The width of the plot in inches
    height = 10) # The height of the plot in inches
d2
e2
r2
dev.off()

```

# Compare gene lists to each other

```{r fig.width=18, warning=FALSE}
# put all gene lists for each DE contrast in one list
compare_list <- list(preVsPost = preVpost_input$gene_list,
                     preVs3hPost = preV3h_input$gene_list)

ck <- compareCluster(geneClusters=compare_list, fun = "GSEA", 
                     TERM2GENE = GO_BP,
                     pvalueCutoff = 0.05,
                     minGSSize = 10, 
                     maxGSSize = 800,
                     eps = 0,
                     by = "fgsea",
                     verbose = FALSE)


dotplot(ck, showCategory=7, split=".sign",
        label_format = 40,
        font.size = 10,) +
        facet_grid(.~.sign) + 
        labs(title = "Gene Set Enrichment Analysis comparison for PreVsPost and PreVs3hPost DE Genes", 
             subtitle = "Top 7 Genes: FDR < 0.05, -1 > logfold2Change > 1 ") +
        theme(plot.title=element_text(hjust=0.5, size = 25, face = "bold"),
              plot.subtitle=element_text(hjust=0.5, size = 20, face = "bold"))
```

### Plot Gene-Concept Network
The cnetplot() depicts the linkages of genes and biological concepts (e.g. GO terms) as a network.
```{r fig.width=18, warning=FALSE}
cnetplot(ck, node_label="category", cex_label_category = 2) + 
  labs(title = "Gene-Concept Network for PreVsPost and PreVs3hPost DE Genes") +
  theme(plot.title=element_text(hjust=0.5, size = 20, face = "bold"))
```

Session Info
```{r session info} 
sessionInfo()
```

